lib/std/c.zig

```diff
@@ -68,324 +69,334 @@ pub fn versionCheck(glibc_version: builtin.Version) type {
-pub extern "c" fn pwritev(fd: c_int, iov: [*]const iovec_const, iovcnt: c_uint, offset: off_t) isize;
-pub extern "c" fn write(fd: fd_t, buf: [*]const u8, nbyte: usize) isize;
-pub extern "c" fn pwrite(fd: fd_t, buf: [*]const u8, nbyte: usize, offset: off_t) isize;
-pub extern "c" fn mmap(addr: ?*align(page_size) c_void, len: usize, prot: c_uint, flags: c_uint, fd: fd_t, offset: off_t) *c_void;
-pub extern "c" fn munmap(addr: *align(page_size) const c_void, len: usize) c_int;
-pub extern "c" fn mprotect(addr: *align(page_size) c_void, len: usize, prot: c_uint) c_int;
-pub extern "c" fn link(oldpath: [*:0]const u8, newpath: [*:0]const u8, flags: c_int) c_int;
-pub extern "c" fn linkat(oldfd: fd_t, oldpath: [*:0]const u8, newfd: fd_t, newpath: [*:0]const u8, flags: c_int) c_int;
-pub extern "c" fn unlink(path: [*:0]const u8) c_int;
-pub extern "c" fn unlinkat(dirfd: fd_t, path: [*:0]const u8, flags: c_uint) c_int;
-pub extern "c" fn getcwd(buf: [*]u8, size: usize) ?[*]u8;
-pub extern "c" fn waitpid(pid: pid_t, stat_loc: ?*c_int, options: c_int) pid_t;
-pub extern "c" fn fork() c_int;
-pub extern "c" fn access(path: [*:0]const u8, mode: c_uint) c_int;
-pub extern "c" fn faccessat(dirfd: fd_t, path: [*:0]const u8, mode: c_uint, flags: c_uint) c_int;
-pub extern "c" fn pipe(fds: *[2]fd_t) c_int;
-pub extern "c" fn mkdir(path: [*:0]const u8, mode: c_uint) c_int;
-pub extern "c" fn mkdirat(dirfd: fd_t, path: [*:0]const u8, mode: u32) c_int;
-pub extern "c" fn symlink(existing: [*:0]const u8, new: [*:0]const u8) c_int;
-pub extern "c" fn symlinkat(oldpath: [*:0]const u8, newdirfd: fd_t, newpath: [*:0]const u8) c_int;
-pub extern "c" fn rename(old: [*:0]const u8, new: [*:0]const u8) c_int;
-pub extern "c" fn renameat(olddirfd: fd_t, old: [*:0]const u8, newdirfd: fd_t, new: [*:0]const u8) c_int;
-pub extern "c" fn chdir(path: [*:0]const u8) c_int;
-pub extern "c" fn fchdir(fd: fd_t) c_int;
-pub extern "c" fn execve(path: [*:0]const u8, argv: [*:null]const ?[*:0]const u8, envp: [*:null]const ?[*:0]const u8) c_int;
-pub extern "c" fn dup(fd: fd_t) c_int;
-pub extern "c" fn dup2(old_fd: fd_t, new_fd: fd_t) c_int;
-pub extern "c" fn readlink(noalias path: [*:0]const u8, noalias buf: [*]u8, bufsize: usize) isize;
-pub extern "c" fn readlinkat(dirfd: fd_t, noalias path: [*:0]const u8, noalias buf: [*]u8, bufsize: usize) isize;
-
 pub usingnamespace switch (builtin.os.tag) {
-    .macos, .ios, .watchos, .tvos => struct {
-        pub const realpath = @"realpath$DARWIN_EXTSN";
-        pub const fstatat = _fstatat;
-    },
+    // WASI-libc should have an empty library_name on extern statements for libc functions
+    // so that we don't have "wasm-import-{name,module}" on these function symbols. Otherwise,
+    // symbols are not resolved by wasm-ld at link time and supposed to be
+    // resolved at load time by runtimes. So instead we define them in "c/wasi.zig".
+    .wasi => struct {},
```

Comment:
I put a comment about why we fail to link wasi-libc with `"c"` or even `"env"` specifier on extern statements.

Reply:
Uh that's weird. I'd expect `env` to work. Have you tired inspecting the LLVM IR and or wasm code?

Reply:
Actually I checked the LLVM IR and Wasm binary (before and after link), and found that 
- `wasm-import-{name,module}` is added only when we provide specifiers on extern statements as I explained in the comment, though this is somewhat expected as I understand from [the code](https://github.com/ziglang/zig/blob/master/src/stage1/codegen.cpp#L454-L456).
- import and function name sections are exactly the same between Wasm binaries with and without "env" specifiers.

Having said that, I guess this has something to do with Link metadata section(https://github.com/WebAssembly/tool-conventions/blob/master/Linking.md#linking-metadata-section) and maybe it is affected by the existence of "wasm-import-*" in the IR.

Reply:
I image `wasm-import-*` would emit some metadata to link_metadata section, and that results in telling wasm-ld not to resolve these symbols at link time.